packages/txservice/src/dispatch.ts

```diff
@@ -0,0 +1,195 @@
+import { GasPrice, WriteTransaction } from "./types";
+import { Transaction } from "./transaction";
+import { TimeoutError, TransactionReverted } from "./error";
+
+// TODO: sep file
+// Thread safe stack manager.
+class TransactionBuffer {
+  // Both of the following are functionally treated as stacks (LIFO).
+  private validated: Transaction[] = [];
+  private pending: Transaction[] = [];
+  // Should be used to access either of the stacks.
+  private readonly accessQueue: PriorityQueue = new PriorityQueue({ concurrency: 1 });
+
+  // Push a new tx to pending.
+  public async push(transaction: Transaction) {
+    // TODO: We may want to cap the pending stack here.
+    await this.accessQueue.add(() => {
+      this.pending.push(transaction);
+    });
+  }
+
+  // Move latest pending to completed.
+  // @returns boolean whether operation was successful, or if there was no pending tx to shift.
+  public async shift(id: string): Promise<boolean> {
+    // TODO: Trim the validated stack when it gets "too long" to save memory space.
+    return await this.accessQueue.add((): boolean => {
+      const transaction = this.pending.find((transaction) => transaction.id === id);
+      if (transaction == null) {
+        return false;
+      }
+      this.validated.push(transaction);
+      return true;
+    });
+  }
+
+  /**
+   * Get the current nonce value.
+   *
+   * @remarks
+   * Caller should still be prepared to get the incorrect nonce back. For instance,
+   * if the provider that just handled our sent tx has suddenly gone offline, this
+   * method may give the wrong nonce. This can be solved by making additional calls to
+   * submit the tx.
+   *
+   * @returns A number value for the current nonce.
+   */
+  public async getNonce(): Promise<number | undefined> {
+    // Update nonce value to greatest of all nonce values retrieved.
+    return (await this.getLastTx())?.nonce;
+  }
+
+  private async getLastTx(): Promise<Transaction | null> {
+    return await this.accessQueue.add((): Transaction | null => {
+      if (this.pending.length > 0) {
+        return this.pending[this.pending.length - 1];
+      } else if (this.validated.length > 0) {
+        return this.validated[this.validated.length - 1];
+      }
+      return null;
+    });
+  }
+}
+
+// TODO: We may want to remodel dispatch to either extend ChainRpcProvider or just move all the ChainRpcProvider functionality over here.
+// For now, a sep. class for simplicity of introducing new functionality.
+/**
+ * @classdesc Dispatches and monitors transaction queue.
+ */
+export class TransactionDispatch {
+  public readonly chainId: number;
+
+  private readonly submitQueue: PriorityQueue = new PriorityQueue({ concurrency: 1 });
+  // A queue for execution of transaction validation. When it completes, we'll shift the transaction from
+  // pending stack -> validated stack in the buffer.
+  private readonly validateQueue: PriorityQueue = new PriorityQueue({ concurrency: 1 });
```

Comment:
We should keep these `submit` / `validate` closer to the nomenclature used when talking about transactions (i.e. `submit` / `mined` / `confirmed`)

Reply:
minedQueue yeah that works

Reply:
Although we should *not* broadcast a tx is mined until after it received the desired # of transactions though so...